File: web/packages/backend/src/modules/orcamento/orcamento.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service.js';
import { OrcamentoFilters } from '@erp-staccato/shared';

@Injectable()
export class OrcamentoService {
  constructor(private prisma: PrismaService) {}

  async list(filters: OrcamentoFilters, userId: number, userType: string, userLojaId: number) {
    try {
      // Build WHERE clause based on filters
      const whereConditions: string[] = [];
      const params: any[] = [];

      // Role-based filtering for vendedor
      if (userType === 'VENDEDOR' || userType === 'VENDEDOR ESPECIAL') {
        if (filters.apenasPropriosOrcamentos) {
          whereConditions.push('(o.idUsuario = ? OR o.idUsuarioConsultor = ?)');
          params.push(userId, userId);
        }
      }

      // Store filter (gerentes see only their store, admin sees all)
      if (userType === 'GERENTE LOJA' || userType === 'GERENTE DEPARTAMENTO') {
        whereConditions.push('o.idLoja = ?');
        params.push(userLojaId);
      } else if (filters.idLoja) {
        whereConditions.push('o.idLoja = ?');
        params.push(filters.idLoja);
      }

      // Status filter
      if (filters.statuses && filters.statuses.length > 0) {
        const placeholders = filters.statuses.map(() => '?').join(',');
        whereConditions.push(`o.status IN (${placeholders})`);
        params.push(...filters.statuses);
      }

      // Month filter (YYYY-MM format)
      if (filters.mesAno) {
        whereConditions.push('o.data2 = ?');
        params.push(filters.mesAno);
      }

      // Vendor filter
      if (filters.idVendedor) {
        whereConditions.push('(o.idUsuario = ? OR o.idUsuarioConsultor = ?)');
        params.push(filters.idVendedor, filters.idVendedor);
      }

      // Supplier filter
      if (filters.fornecedor) {
        whereConditions.push('FIND_IN_SET(?, o.fornecedores)');
        params.push(filters.fornecedor);
      }

      // Search filter (idOrcamento, cliente, profissional, vendedor)
      if (filters.search) {
        const searchTerm = `%${filters.search}%`;
        whereConditions.push('(o.idOrcamento LIKE ? OR c.nome_razao LIKE ? OR p.nome_razao LIKE ? OR u.nome LIKE ?)');
        params.push(searchTerm, searchTerm, searchTerm, searchTerm);
      }

      const whereClause = whereConditions.length > 0 ? `WHERE ${whereConditions.join(' AND ')}` : '';

      // Use raw query to access view_orcamento which has computed fields
      const orcamentos = await this.prisma.$queryRawUnsafe(
        `
        SELECT
          o.idOrcamento,
          o.idLoja,
          o.idUsuario,
          o.idUsuarioConsultor,
          o.status,
          CASE
            WHEN o.status = 'FECHADO' THEN ''
            WHEN o.status = 'PERDIDO' THEN ''
            WHEN o.status = 'CANCELADO' THEN ''
            WHEN DATEDIFF(DATE_ADD(o.data, INTERVAL o.validade DAY), CURDATE()) < 0 THEN 'EXPIRADO'
            ELSE DATEDIFF(DATE_ADD(o.data, INTERVAL o.validade DAY), CURDATE())
          END as diasRestantes,
          u.nome as vendedor,
          uc.nome as consultor,
          c.nome_razao as cliente,
          p.nome_razao as profissional,
          c.tel,
          c.telCel,
          p.tel as telProf,
          o.data,
          o.data2,
          o.total,
          o.created,
          o.lastUpdated,
          o.observacao,
          o.fornecedores
        FROM orcamento o
        LEFT JOIN usuario u ON o.idUsuario = u.idUsuario
        LEFT JOIN usuario uc ON o.idUsuarioConsultor = uc.idUsuario
        LEFT JOIN cliente c ON o.idCliente = c.idCliente
        LEFT JOIN profissional p ON o.idProfissional = p.idProfissional
        ${whereClause}
        ORDER BY o.data DESC
        `,
        ...params
      );

      return orcamentos || [];
    } catch (error) {
      console.error('Error listing orcamentos:', error);
      throw error;
    }
  }

  async getLojasForFilter() {
    return this.prisma.loja.findMany({
      where: { desativado: false },
      select: { idLoja: true, descricao: true, nomeFantasia: true },
      orderBy: { descricao: 'asc' },
    });
  }

  async getVendedoresForFilter(idLoja?: number) {
    return this.prisma.usuario.findMany({
      where: {
        tipo: { in: ['VENDEDOR', 'VENDEDOR ESPECIAL'] },
        desativado: false,
        ...(idLoja && { idLoja }),
      },
      select: { idUsuario: true, nome: true },
      orderBy: { nome: 'asc' },
    });
  }

  async getFornecedoresForFilter() {
    // Extract unique fornecedores from orcamentos
    const result = await this.prisma.$queryRaw<{ fornecedores: string | null }[]>`
      SELECT DISTINCT fornecedores FROM orcamento
      WHERE fornecedores IS NOT NULL AND fornecedores != ''
    `;

    const fornecedoresSet = new Set<string>();
    result.forEach((row: { fornecedores: string | null }) => {
      if (row.fornecedores) {
        row.fornecedores.split(',').forEach((f: string) => {
          fornecedoresSet.add(f.trim());
        });
      }
    });

    return Array.from(fornecedoresSet)
      .sort()
      .map((name) => ({ razaoSocial: name }));
  }
}

File: web-dotnet/src/Backend/Services/OrcamentoService.cs
using Microsoft.EntityFrameworkCore;
using ERP.Staccato.Backend.Data;
using ERP.Staccato.Shared.Models;

namespace ERP.Staccato.Backend.Services;

/// <summary>
/// Orcamento (Budget) service - mirrors TypeScript orcamento.service.ts
/// </summary>
public class OrcamentoService
{
    private readonly ApplicationDbContext _context;

    public OrcamentoService(ApplicationDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// List orcamentos with filtering
    /// Mirrors C++ WidgetOrcamento::montaFiltro() logic
    /// Implemented using LINQ for compatibility with both SQL Server and in-memory testing
    /// </summary>
    public async Task<List<OrcamentoListItem>> ListAsync(
        OrcamentoFilters filters,
        int userId,
        string userType,
        int userLojaId,
        string userName)
    {
        try
        {
            var query = _context.Set<OrcamentoView>().AsQueryable();

            // Role-based filtering
            if (userType == "GERENTE LOJA" || userType == "GERENTE DEPARTAMENTO")
            {
                query = query.Where(o => o.IdLoja == userLojaId);
            }
            else if (filters.IdLoja.HasValue)
            {
                query = query.Where(o => o.IdLoja == filters.IdLoja);
            }

            // Month filter
            if (!string.IsNullOrEmpty(filters.MesAno))
            {
                query = query.Where(o => o.Data2 == filters.MesAno);
            }

            // Vendor filter
            if (filters.IdVendedor.HasValue)
            {
                query = query.Where(o =>
                    o.IdUsuario == filters.IdVendedor ||
                    o.IdUsuarioConsultor == filters.IdVendedor);
            }

            // Supplier filter
            if (!string.IsNullOrEmpty(filters.Fornecedor))
            {
                query = query.Where(o =>
                    o.Fornecedores != null &&
                    o.Fornecedores.Contains(filters.Fornecedor));
            }

            // Status filter
            if (filters.Statuses.Any())
            {
                query = query.Where(o => filters.Statuses.Contains(o.Status));
            }

            // Semaforo filter
            if (filters.Semaforo.HasValue)
            {
                query = query.Where(o => o.Semaforo == filters.Semaforo);
            }

            // "Próprios" filter (only for vendedores)
            if ((userType == "VENDEDOR" || userType == "VENDEDOR ESPECIAL")
                && filters.ApenasPropriosOrcamentos)
            {
                query = query.Where(o =>
                    o.Vendedor == userName ||
                    o.Consultor == userName);
            }

            // Search filter
            if (!string.IsNullOrEmpty(filters.Search))
            {
                var searchTerm = filters.Search;
                query = query.Where(o =>
                    o.IdOrcamento.ToString().Contains(searchTerm) ||
                    (o.Vendedor != null && o.Vendedor.Contains(searchTerm)) ||
                    (o.Cliente != null && o.Cliente.Contains(searchTerm)) ||
                    (o.Profissional != null && o.Profissional.Contains(searchTerm))
                );
            }

            // Order by
            query = query.OrderByDescending(o => o.Data);

            Console.WriteLine($"[Orcamento.ListAsync] Filters: {System.Text.Json.JsonSerializer.Serialize(filters)}");

            // Execute query and map results
            var orcamentos = await query.ToListAsync();

            var result = orcamentos.Select(o => new OrcamentoListItem
            {
                IdOrcamento = o.IdOrcamento,
                IdLoja = o.IdLoja,
                IdUsuario = o.IdUsuario,
                IdUsuarioConsultor = o.IdUsuarioConsultor,
                Status = o.Status,
                DiasRestantes = o.DiasRestantes?.ToString() ?? "-",
                Vendedor = o.Vendedor,
                Consultor = o.Consultor,
                Cliente = o.Cliente,
                Profissional = o.Profissional,
                Tel = o.Tel,
                TelCel = o.TelCel,
                TelProf = o.TelProf,
                Data = o.Data,
                Data2 = o.Data2,
                Total = o.Total,
                DataFollowup = o.DataFollowup,
                DataProxFollowup = o.DataProxFollowup,
                Observacao = o.Observacao,
                Semaforo = o.Semaforo,
                Fornecedores = o.Fornecedores
            }).ToList();

            Console.WriteLine($"[Orcamento.ListAsync] Returned {result.Count} rows");

            return result;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error listing orcamentos: {ex}");
            throw;
        }
    }

    /// <summary>
    /// Get lojas for filter dropdown
    /// </summary>
    public async Task<List<LojaDto>> GetLojasForFilterAsync()
    {
        return await _context.Set<Loja>()
            .Where(l => !l.Desativado)
            .OrderBy(l => l.IdLoja)
            .Select(l => new LojaDto
            {
                IdLoja = l.IdLoja,
                Descricao = l.Descricao,
                NomeFantasia = l.NomeFantasia
            })
            .ToListAsync();
    }

    /// <summary>
    /// Get vendedores for filter dropdown
    /// </summary>
    public async Task<List<VendedorDto>> GetVendedoresForFilterAsync(int? idLoja = null)
    {
        var query = _context.Set<Usuario>()
            .Where(u => (u.Tipo == "VENDEDOR" || u.Tipo == "VENDEDOR ESPECIAL") && !u.Desativado);

        if (idLoja.HasValue)
            query = query.Where(u => u.IdLoja == idLoja);

        return await query
            .OrderBy(u => u.Nome)
            .Select(u => new VendedorDto
            {
                IdUsuario = u.IdUsuario,
                Nome = u.Nome
            })
            .ToListAsync();
    }

    /// <summary>
    /// Get fornecedores for filter dropdown
    /// </summary>
    public async Task<List<FornecedorDto>> GetFornecedoresForFilterAsync()
    {
        // Get all orcamentos with fornecedores data
        var orcamentos = await _context.Set<OrcamentoView>()
            .Where(o => o.Fornecedores != null && o.Fornecedores != "")
            .Select(o => o.Fornecedores)
            .ToListAsync();

        var fornecedoresSet = new HashSet<string>();

        foreach (var fornecedoresStr in orcamentos)
        {
            if (!string.IsNullOrEmpty(fornecedoresStr))
            {
                var fornecedores = fornecedoresStr.Split(',');
                foreach (var f in fornecedores)
                {
                    var trimmed = f.Trim();
                    if (!string.IsNullOrEmpty(trimmed))
                        fornecedoresSet.Add(trimmed);
                }
            }
        }

        return fornecedoresSet
            .OrderBy(x => x)
            .Select(x => new FornecedorDto { RazaoSocial = x })
            .ToList();
    }
}
